service/matching/taskQueueManager.go

```diff
@@ -426,16 +453,59 @@ func (c *taskQueueManagerImpl) DispatchQueryTask(
 	return c.matcher.OfferQuery(ctx, task)
 }
 
+// GetVersioningData returns the versioning data for the task queue if any. If this task queue is a sub-partition and
+// has no cached data, it will explicitly attempt a fetch from the root partition.
 func (c *taskQueueManagerImpl) GetVersioningData(ctx context.Context) (*persistencespb.VersioningData, error) {
-	return c.db.GetVersioningData(ctx)
+	vd, err := c.db.GetVersioningData(ctx)
+	if errors.Is(err, &versioningDataNotPresentOnSubPartition{}) {
+		// If this is a sub-partition with no versioning data, this call is indicating we might expect to find
+		// some. Since we may not have started the poller, we want to explicitly attempt to fetch now, because
+		// it's possible some was added to the root partition, but it failed to invalidate this sub partition.
+		return c.fetchMetadataFromRootPartition(ctx)
+	}
+	return vd, err
 }
 
 func (c *taskQueueManagerImpl) MutateVersioningData(ctx context.Context, mutator func(*persistencespb.VersioningData) error) error {
 	err := c.db.MutateVersioningData(ctx, mutator)
+	// If this is the root partition, notify sub partitions that they should fetch changed data from us.
+	if c.taskQueueID.IsRoot() {
+		numParts := c.config.NumReadPartitions()
+		for i := 1; i <= numParts; i++ {
+			tq := c.taskQueueID.mkName(i)
+			_, err := c.matchingClient.InvalidateTaskQueueMetadata(ctx,
+				&matchingservice.InvalidateTaskQueueMetadataRequest{
+					NamespaceId:    c.taskQueueID.namespaceID.String(),
+					TaskQueue:      tq,
+					VersioningData: true,
+				})
+			if err != nil {
+				c.logger.Warn("Failed to notify sub-partition of invalidated versioning data",
+					tag.WorkflowTaskQueueName(tq), tag.Error(err))
+			}
+		}
+	}
 	c.signalIfFatal(err)
 	return err
 }
 
+func (c *taskQueueManagerImpl) InvalidateMetadata(ctx context.Context, request *matchingservice.InvalidateTaskQueueMetadataRequest) error {
+	if request.GetVersioningData() {
+		if c.taskQueueID.IsRoot() {
+			// Should never happen. Root partitions do not get their versioning data invalidated.
+			c.logger.Warn("A root partition was told to invalidate its versioning data, this should not happen",
+				tag.WorkflowTaskQueueName(c.taskQueueID.name))
+			return nil
+		}
+		c.db.setVersioningDataForNonRootPartition(nil)
```

Comment:
why set to nil first? why not just replace so you go from one populated value to another?

what should the matcher do if it needs version info between this line and the next?

Reply:
I think this is something leftover that I didn't need any more. Good catch. Removed.